pkg/fake/subscriptionsapi.go

```diff
@@ -0,0 +1,94 @@
+/*
+Portions Copyright (c) Microsoft Corporation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fake
+
+import (
+	"context"
+	"sort"
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armsubscriptions"
+	"github.com/Azure/karpenter-provider-azure/pkg/providers/zone"
+	"github.com/samber/lo"
+)
+
+type ListLocationsInput struct {
+	SubscriptionID string
+	Options        *armsubscriptions.ClientListLocationsOptions
+}
+
+type SubscriptionsAPIBehavior struct {
+	NewListLocationsPagerBehavior MockedFunction[ListLocationsInput, armsubscriptions.ClientListLocationsResponse]
+	Locations                     sync.Map
+}
+
+var _ zone.SubscriptionsAPI = &SubscriptionsAPI{}
+
+type SubscriptionsAPI struct {
+	SubscriptionsAPIBehavior
+}
+
+func (api *SubscriptionsAPI) Reset() {
+	api.NewListLocationsPagerBehavior.Reset()
+	api.Locations.Range(func(k, v any) bool {
+		api.Locations.Delete(k)
+		return true
+	})
+}
+
+func (api *SubscriptionsAPI) NewListLocationsPager(
+	subscriptionID string,
+	options *armsubscriptions.ClientListLocationsOptions,
+) *runtime.Pager[armsubscriptions.ClientListLocationsResponse] {
+	input := &ListLocationsInput{
+		SubscriptionID: subscriptionID,
+		Options:        options,
+	}
+
+	pagingHandler := runtime.PagingHandler[armsubscriptions.ClientListLocationsResponse]{
+		More: func(page armsubscriptions.ClientListLocationsResponse) bool {
+			return false // TODO: It might be ideal if we had a MockPager which sometimes simulated multiple pages of results to ensure we handle that correctly
+		},
+		Fetcher: func(ctx context.Context, _ *armsubscriptions.ClientListLocationsResponse) (armsubscriptions.ClientListLocationsResponse, error) {
+			return api.NewListLocationsPagerBehavior.Invoke(input, func(input *ListLocationsInput) (armsubscriptions.ClientListLocationsResponse, error) {
```

Comment:
This is a bit weird but I wanted a way to check if we'd called the Fetcher and this seemed the best way with what we ahve.